3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

Yo0oN opened the pull request.

Yo0oN committed:
feat:firebase 적용

1차 작성 완료
이후 send부분 분리, 로그인 후 토큰 생성, 토큰 추가 예정

[#67]feat:3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

Yo0oN committed:
feat:firebase 적용

3개월 간 비밀번호 미변경 시 알림

[#67]feat:3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

Yo0oN commented on src/main/java/com/festa/common/firebase/FirebaseTokenManager.java:
```diff
@@ -0,0 +1,64 @@
+    private String firebaseConfigPath;
+
+    // 토큰을 저장해둘 곳
+    private Map<Long, String> tokenMap = new HashMap<>();
```
일단 HashMap을 이용해서 userNo를 키로, Token을 값으로 저장중인데 저장방식을 바꾸려고 생각중입니다! 지금 레디스에 로그인을하면 세션아이디를 키로, userNo를 값으로 넣고있는데 값을 넣을때 배열을 이용해서 userNo와 Token을 같이 넣어도 괜찮을지, 다른 방법은 없을지 궁금합니다!

f-lab-dev replied:
1. 여기에 토큰을 저장하면 스케일아웃을 했을때 어떻게 될까요? 
2. 토큰이 점점 많이 쌓일 텐데 메모리가 부족해지지 않을까요?
3. 질문에 대한 답을 드리자면 이 데이터는 꼭 세션 데이터에 저장해야하는걸까요?

Yo0oN replied:
1. 각 서버의 HashMap에 들어있는 토큰값이 달라 데이터불일치가 일어납니다.
2. 사용자가 많아지면 메모리가 부족해지고, 데이터불일치 문제도 있기 때문에 세션에 저장해보면 어떨까라는 생각을 해보고있었습니다.
3. 세션과 같은 주기로 토큰이 살아있기 때문에 세션에 함께 저장하는건 어떨까했습니다! 그 외에는 쿠키에저장해보는건 어떨까 하는 생각도 해보았습니다!

f-lab-dev replied:
레디스같은곳에 저장해두면 확장성이 좋아지지 않을까요?

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseAlertMessaageTitle.java:
```diff
@@ -0,0 +1,8 @@
+
+@Component
+public class FirebaseAlertMessaageTitle {
+    public final static String changePwTitle = "비밀번호를 변경해주세요.";
```
상수는 이름을 대문자로 지어주세요~

Yo0oN replied:
네 수정하겠습니다

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseAlertMessageContents.java:
```diff
@@ -0,0 +1,8 @@
+
+import org.springframework.stereotype.Component;
+
+@Component
```
상수를 저장하는 클래스인데 bean으로 생성할 필요가 있을까요?

Yo0oN replied:
이부분은 습관적으로 붙이게 되었습니다. 수정하겠습니다.

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseTokenManager.java:
```diff
@@ -0,0 +1,64 @@
+     * @param token
+     */
+    public void register(long userNo, String token) {
+        tokenMap.put(userNo, token);
```
이 행위를 비롯하여 아래의 메소드들은 전부 thread safe하지 않습니다. 왜 그럴까요?

Jane096 replied:
`HashMap` 자체가 이미 thread safe 하지 않기 때문에 수많은 사용자가 로그인 하게 되면 그 토큰을 매번 저장하게 될 텐데 저장을 하면서 값이 꼬일 가능성이 있다고 생각합니다.

Yo0oN replied:
HashMap은 Thread-safe 하지 않은 클래스이기 때문에 사용하려면 동기화처리를 해주거나, Thread-safe한 ConccurentHashMap을 사용하는것이 좋습니다. 이부분은 위의 토큰 저장하는곳을 어떻게 할지 좀 더 생각해보고 함께 수정하겠습니다.

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseTokenManager.java:
```diff
@@ -0,0 +1,64 @@
+                    .createScoped(Arrays.asList("https://www.googleapis.com/auth/firebase.remoteconfig"));
+            googleCredential.refreshToken();
+        } catch (IOException e) {
+
```
예외를 무시하도록 처리하셨는데 이 것은 무시해도 되는 연산일까요? 아래의 코드에서 NPE가 날 것 같습니다

Jane096 replied:
catch 로 예외를 잡고 아무것도 처리를 안해주면 어떤 예외가 발생했는지 알 길이 없어 지양해야하는 코드라고 봤습니다. catch 절 안에 예외발생 시 무언가 처리해야하는 로직이 필요할 것 같습니다.

Yo0oN replied:
token을 제대로 만들지 못하게 되었다는 예외를 만들어서 추가해주었습니다.

f-lab-dev commented on src/main/java/com/festa/service/AlertService.java:
```diff
@@ -1,16 +1,63 @@
+        String overChangePwDate = alertDAO.sendChangePwNotice(userNo);
+
+        // 변경일이 180이 넘었다면 알림
+        if (overChangePwDate.equals("1")) {
```
`overChangePwDate`가 null일 경우 NPE가 나는 것을 방지하기 위해 상수를 앞에 두는게 낫습니다. `"1".equals(~~)`

StringUtils.equals를 사용하는것도 좋고요

Yo0oN replied:
넵 overChangePwDate를 뒤로 보내는것으로 수정하였습니다!

f-lab-dev commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -126,6 +130,10 @@
 
+        firebaseTokenManager.makeAccessToken(memberDTO.getUserNo());
+
+        alertService.sendChangePwNotice(memberDTO.getUserNo());
```
보통 푸시로 알림을 보내기보다는 홈페이지에서 변경 유도창을 띄워주는데 그렇게 하려면 응답값에다 flag를 주는게 낫지 않을까요?

Yo0oN replied:
파이어베이스로 푸시알림을 보내지 않고 비밀번호 변경 여부를 확인한 후 flag를 주고 프론트에서 처리하도록 하는게 낫다는뜻인가요..?

f-lab-dev replied:
네 맞습니다~ 비밀번호를 변경할 때가 되면 일반적으로 로그인했을때 변경유도창을 띄우죠.

이건 기획에도 문제가 있었던 것 같긴한데 푸시알림을 주려면 주기적으로 변경기한이 지난 사람들에게 푸시를 날려야겠죠~
근데 그렇게 비밀번호 변경 푸시를 자주 날리게 되면 사용자들이 어떻게 반응할지도 서비스적인 측면에서 고민해봐야합니다.

Jane096 replied:
그럼 이전처럼 flag값을 날려서 true 일 경우에만 변경할 수 있도록 하는게 서비스 측면에서 고려해봤을 때 좋은 방법일 것 같습니다. 실제 서비스도 그렇게 돌아가니까요. @Yo0oN

Jane096 commented on src/main/java/com/festa/dto/AlertDTO.java:
```diff
@@ -9,11 +9,13 @@
 @Builder
 public class AlertDTO {
 
-    long userNo;
+    int userNo;
```
`int` 는 범위가 좁다는 생각이 들어서 저는 `userNo` 를 모두 `long`  으로 변경했습니다!

Yo0oN replied:
네 수정했습니다!

Jane096 commented on src/main/java/com/festa/common/firebase/FirebaseInitializer.java:
```diff
@@ -0,0 +1,48 @@
+import java.io.FileInputStream;
+import java.io.IOException;
+
+@Service
+public class FirebaseInitializer {
+    @Value("${firebase.firebaseConfigPath}")
+    private String firebaseConfigPath;
```
클래스 선언부와 `@Value`  선언 사이는 한 줄 띄어주는 게 좋을 것 같습니다!

Yo0oN committed:
feat:3개월 간 비밀번호 미변경 시 알림

푸시 알림 대신 return값에 flag를 담아주어 true일경우 프론트에서 알림창 띄우도록 수정
HashMap 대신 thread-safe한 ConccurentHashMap으로 수정

[#67]feat:3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

f-lab-dev commented on src/main/java/com/festa/service/MemberService.java:
```diff
@@ -38,4 +38,10 @@ public void memberWithdraw(MemberDTO memberDTO) {
+    public boolean getChangePwDateDiff(long userNo) {
+        String overChangePwDate = memberDAO.getChangePwDateDiff(userNo);
+
+        return "1".equals(overChangePwDate) ? true : false;
```
이렇게 할 필요없이 mysql에서 boolean을 리턴하게할 방법이 있지 않을까요?

Jane096 replied:
제가 생각해보았을 때 

```xml
<select id="sendChangePwNotice" resultType="boolean">
        SELECT
            if(DATEDIFF(NOW(), pwLastModified) > 90, true, false) AS sendAlert
        FROM notice
        WHERE userNo = #{userNo}
    </select>
```

이렇게 하면 `boolean`을 리턴하게 할 수 있을 것 같은데 
다른 분들의 의견 궁금합니다!

Jane096 replied:
```xml
<select id="sendChangePwNotice" resultType="boolean">
```

"String" 이 아니라 "boolean" 입니당.. 오타있었네요

f-lab-dev replied:
이런 것을 리뷰할땐 답을 알려주기보다는 찾는법을 알려주어야합니다~
리뷰어는 리뷰를 받는 사람이 리뷰로 인해 더 성장할 수 있게 가이드를 하는 역할도 하기 때문입니다

Jane096 replied:
제가 너무 앞서갔던 것 같네요..  리뷰달 때 주의하도록 하겠습니다..!

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseTokenManager.java:
```diff
@@ -9,14 +10,16 @@
 
     // 토큰을 저장해둘 곳
-    private Map<Long, String> tokenMap = new HashMap<>();
+    private Map<Long, String> tokenMap = new ConcurrentHashMap<>();
```
여기선 토큰을 WAS의 메모리에 저장하고 있습니다. 이렇게 하면 어떤 문제가 발생하는지 이전 기능을 개발하면서 공부했었던 것 같네요

Jane096 commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -139,6 +145,8 @@
     public ResponseEntity<HttpStatus> logout() {
         loginService.removeUserNo();
 
+        firebaseTokenManager.removeToken(userNo);
+
```
이 부분에서 `userNo` 파라미터를 받는 부분이 없어서 에러가 발생하고 있습니다!

Jane096 commented on src/main/java/com/festa/common/firebase/FirebaseInitializer.java:
```diff
@@ -0,0 +1,49 @@
+    public void firebaseInit() {
+        try {
+            FileInputStream serviceAccount = new FileInputStream(firebaseConfigPath);
+
+            FirebaseOptions options = new FirebaseOptions.Builder()
```
`Builder()` 라는 메서드는 현재 `deprecated` 된 기능인 것 같습니다! 다른 대체 메서드가 있는 것 같은데 그걸로 변경이 필요할 것 같습니다~

Yo0oN replied:
네~ deprecated되지 않은것으로 변경하였습니다!

Jane096 reviewed (changes requested):
approve 다시 해제하겠습니다...!!

Yo0oN committed:
feat:3개월 간 비밀번호 미변경 시 알림

FirebaseOptions.Builder()대신 deprecated되지 않은 builder()로 수정
ConcurrentHashMap에 저장중이던 토큰을 Redis에 저장
MemberDAO.getChangePwDateDiff 리턴값 boolean으로 변경

[#67]feat:3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseTokenManager.java:
```diff
@@ -2,49 +2,49 @@
-    public void register(long userNo, String token) {
-        tokenMap.put(userNo, token);
+    public void register(String userNo, String token) {
+        httpSession.setAttribute(userNo, token);
```
레디스에 저장한다와 세션에 저장한다는 다른 의미입니다.
세션에 저장하면 만료시간이 너무 길어지지 않을까요?

Yo0oN replied:
수정하였습니다.!

Yo0oN committed:
feat:3개월 간 비밀번호 미변경 시 알림

토큰 Redis에 저장하도록 수정
deprecated된 GoogleCredential대신 GoogleCredentials로 변경

[#67]feat:3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

Yo0oN committed:
feat:3개월 간 비밀번호 미변경 시 알림

MemberController에서 더이상 사용하지 않는 AlertService 삭제

[#67]feat:3개월 간 비밀번호 미변경 시 알림서비스 기능 수정

f-lab-dev commented on src/main/java/com/festa/common/firebase/FirebaseTokenManager.java:
```diff
@@ -27,7 +27,7 @@
      */
     public void register(String userNo, String token) {
-        httpSession.setAttribute(userNo, token);
+        redisTemplate.opsForValue().set(userNo, token);
```
이렇게 하면 만료시간은 얼마가 되는걸까요?

Yo0oN replied:
기본적으로 토큰 만료시간은 한시간이고, 추가 설정을 해주면 만료시간을 조정할 수 있습니다. 지금은 기본값이라 1시간이며, 로그아웃할때 레디스에서 제거하도록 해두었습니다!

Yo0oN committed:
Merge branch 'develop' into feature/#67
